src/BufferedRequest.ts

```diff
@@ -0,0 +1,38 @@
+import { Headers } from 'headers-polyfill/lib'
+import { decodeBuf } from './utils/bufferCodec'
+import { uuidv4 } from './utils/uuid'
+
+export class BufferedRequest {
+  public id = uuidv4()
+
+  constructor(
+    public readonly url: URL,
+    private readonly body: ArrayBuffer,
+    private readonly init: RequestInit
+  ) {}
+
+  public get method(): string {
+    return this.init.method || 'GET'
+  }
+
+  public text(): string {
+    return decodeBuf(this.body)
+  }
+
+  public json<T = any>(): T {
```

Comment:
To keep compatibility with Fetch API Request, let's make `text`, `json`, and `arrayBuffer` return a Promise that resolves the values you have now.

Reply:
Fetch api has those for response. also, currently most of my mocks are synchronous. Making these return promise will force the user to use async await.

Reply:
It looks like it's the same for Request:

- [`Request.arrayBuffer()`](https://developer.mozilla.org/en-US/docs/Web/API/Request/arrayBuffer)
- [`Request.json()`](https://developer.mozilla.org/en-US/docs/Web/API/Request/json)
- [`Request.text()`](https://developer.mozilla.org/en-US/docs/Web/API/Request/text)

Those methods return promises. Let's keep our abstraction compatible. 

I'm advocating for fetch API compatibility because if we achieve that, we will be able to use an intercepted request instance as an input to regular `window.fetch()`, which is a great way to perform [response patching](https://mswjs.io/docs/recipes/response-patching).

Reply:
Also, an interesting detail from the docs:

> The response is always decoded using UTF-8.
> — Request.text()

To the point of `text()` accepting a custom encoding. Turns out it can't.